query_optimizer/OptimizerTree.hpp

```diff
@@ -118,6 +122,169 @@ class OptimizerTree : public OptimizerTreeBase {
+        : key_(key), value_(value) {  }
+
+    std::string getName() const {
+      return "ProtoProperty";
+    }
+
+    void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                             std::vector<std::string> *inline_field_values,
+                             std::vector<std::string> *non_container_child_field_names,
+                             std::vector<TreeNodeType> *non_container_child_fields,
+                             std::vector<std::string> *container_child_field_names,
+                             std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+      inline_field_names->push_back(std::string("Property"));
+      inline_field_values->push_back(key_);
+
+      inline_field_names->push_back(std::string("Value"));
+      inline_field_values->push_back(value_);
+    }
+
+   private:
+    std::string key_;
+    std::string value_;
+  };  // class OptimizerProtoRepresentation
+
+ public:
+  OptimizerProtoRepresentation() : properties_() { }
+
+  std::string getName() const {
+    return "ProtoDescription";
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value A string value.
+   */
+  void addProperty(std::string key, std::string value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, value));
+    properties_.push_back(property);
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value An int value which is converted into a string.
+   */
+  void addProperty(std::string key, int value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, std::to_string(value)));
+    properties_.push_back(property);
+  }
+
+  void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                           std::vector<std::string> *inline_field_values,
+                           std::vector<std::string> *non_container_child_field_names,
+                           std::vector<TreeNodeType> *non_container_child_fields,
+                           std::vector<std::string> *container_child_field_names,
+                           std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+    for (auto property : properties_) {
+      non_container_child_field_names->push_back("Property");
+      non_container_child_fields->push_back(property);
+    }
+  }
+  // A list of managed properties.
+  std::vector<std::shared_ptr<const OptimizerProtoPropertyRepresentation> > properties_;
```

Comment:
Yes, I will add a comment. In a nutshell, when the vector goes out of scope, so do its contained unique ptrs. All of the unique pointers will call the dtors of their managed objects:

```c++
#include <iostream>
#include <string>
#include <memory>
#include <vector>

using namespace std;

class B {
 public:
  B(const char* s) : name(s) {}
  ~B() {
      cout << name << " has been deleted.\n";
  }
  string name; 
};

int main() {
	{
	    vector<unique_ptr<B> > vec;
	    unique_ptr<B> pb1(new B("bob"));
	    vec.push_back(std::move(pb1));
	    unique_ptr<B> pb2(new B("fred"));
	    vec.push_back(std::move(pb2));
	    unique_ptr<B> pb3(new B("chuck"));
	    vec.push_back(std::move(pb3));
	}
	return 0;
}
```

Will print:
```
bob has been deleted.
fred has been deleted.
chuck has been deleted.
```